core/src/main/java/amino/run/kernel/client/KernelClient.java

```diff
@@ -28,41 +34,180 @@
+                InetSocketAddress serverAddress, KernelServer remoteRef, final NodeMetric metric) {
+            this.serverAddress = serverAddress;
+            this.remoteRef = remoteRef;
+            this.metric = metric;
+            data = new RandomData();
+            metricsTimer =
+                    new ResettableTimer(
+                            new TimerTask() {
+                                public void run() {
+                                    measureMetrics(KernelServerInfo.this);
+                                }
+                            },
+                            metricPollPeriod);
+            metricsTimer.start();
+        }
+    }
+
+    /** Map of remote kernel server hostname to their kernel server info */
+    private ConcurrentHashMap<InetSocketAddress, KernelServerInfo> servers;
+
+    /* Map of remote kernel server hostname to their metrics. It is maintained independent from servers map, so that
+    this map can be serialized and sent to OMS as is */
+    private ConcurrentHashMap<InetSocketAddress, NodeMetric> metrics;
 
     private static final Logger logger = Logger.getLogger(KernelClient.class.getName());
 
+    public ConcurrentHashMap<InetSocketAddress, NodeMetric> getMetrics() {
+        return metrics;
+    }
+
     /**
-     * Add a host to the list of hosts that we've contacted
+     * Adds remote kernel server information to cache for the given host
      *
      * @param host
      */
-    private KernelServer addHost(InetSocketAddress host) {
+    private KernelServerInfo addHost(InetSocketAddress host) {
         try {
             Registry registry = LocateRegistry.getRegistry(host.getHostName(), host.getPort());
             KernelServer server = (KernelServer) registry.lookup("io.amino.run.kernelserver");
-            servers.put(host, server);
-            return server;
+            NodeMetric metric = new NodeMetric();
+            KernelServerInfo serverInfo = new KernelServerInfo(host, server, metric);
+            servers.put(host, serverInfo);
+            metrics.put(host, metric);
+            return serverInfo;
         } catch (Exception e) {
-            logger.severe("Could not find MicroService server on host: " + e.toString());
+            logger.severe(
+                    String.format(
+                            "Could not find kernel server on host: %s. Exception: %s", host, e));
         }
         return null;
     }
 
+    /**
+     * Removes cached remote kernel server information for the given host, if it has missed
+     * heartbeats for {@link amino.run.kernel.client.KernelClient#MAX_HEARTBEAT_MISS_TIMES} times
+     *
+     * @param host
+     */
+    private void removeHost(InetSocketAddress host) {
+        KernelServerInfo server = servers.get(host);
+        if (server != null) {
+            if ((server.refCount - server.refCountOnHeartBeat >= MAX_HEARTBEAT_MISS_TIMES)) {
+                servers.remove(host);
+                metrics.remove(host);
```

Comment:
See above comment.  Merging would remove this duplication.

Reply:
Modified